### PR TITLE
Generate deployment plan

### PR DESCRIPTION
**Pull request by moelsayed**
https://github.com/rancher/types/pull/206

### CODE REVIEW

File: cluster/plan.go [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+		PidMode:       "host",
+		Privileged:    true,
+		RestartPolicy: container.RestartPolicy{Name: "always"},
+	}
+	for arg, value := range c.Services.Kubelet.ExtraArgs {
+		cmd := fmt.Sprintf("--%s=%s", arg, value)
+		imageCfg.Entrypoint = append(imageCfg.Entrypoint, cmd)
+	}
+	healthCheck := v3.HealthCheck{
+		URL: fmt.Sprintf("%s%s:%d%s", services.HTTPSProtoPrefix, services.HealthzAddress, services.KubeletPort, services.HealthzEndpoint),
```
Comment (galal-hussein) [inline cluster/plan.go]: can you create one unified function to return the healthcheck URL and send parameters to it

File: cluster/network.go [hunk 2]
```diff
@@ -41,6 +41,9 @@ const (
 	KubeProxyPort       = "10256"
 	FlannetVXLANPortUDP = "8472"
 
+	ProtocolTCP = "TCP"
+	ProtocolUDP = "UDP"
```
Comment (galal-hussein) [inline cluster/network.go]: Is this used somewhere?

File: services/workerplane.go [hunk 3]
```diff
@@ -13,18 +13,23 @@ const (
...
-	allHosts := hosts.GetUniqueHostList(etcdHosts, controlHosts, workerHosts)
 	for _, host := range allHosts {
 		if !host.IsControl && !host.IsWorker {
 			// Add unschedulable taint
 			host.ToAddTaints = append(host.ToAddTaints, unschedulableEtcdTaint)
 		}
 		runHost := host
+		tmpMap := make(map[string]v3.Process)
+		// maps are not thread safe
+		for k, v := range processMap {
```
Comment (galal-hussein) [inline services/workerplane.go]: Can copy map be moved to a separate function

File: cluster/plan.go [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+			"/etc/kubernetes:/etc/kubernetes:z",
+		},
+		NetworkMode: "host",
+	}
+	for arg, value := range c.Services.Etcd.ExtraArgs {
+		cmd := fmt.Sprintf("--%s=%s", arg, value)
+		imageCfg.Entrypoint = append(imageCfg.Entrypoint, cmd)
+	}
+	healthCheck := v3.HealthCheck{
+		URL: "https://127.0.0.1:2379/health",
```
Comment (galal-hussein) [inline cluster/plan.go]: Can you add this to a constant

File: cluster/network.go (hunk 2 above)
Comment (moelsayed) [reply cluster/network.go]: Will be used later to support UDP checks.

File: cluster/cluster.go [hunk 5]
```diff
@@ -284,3 +297,11 @@ func ConfigureCluster(ctx context.Context, rkeConfig v3.RancherKubernetesEngineC
 
 	return kubeCluster.deployAddons(ctx)
 }
+
+func (c *Cluster) getEtcdProcessHostMap(etcdHosts []*hosts.Host) map[*hosts.Host]v3.Process {
```
Comment (galal-hussein) [inline cluster/cluster.go]: Can this parameter be more descriptive like `readyEtcdHosts` or you may add a comment explaining the logic of the function, thanks!

**Pull request merged**